tmk_core/common/action_util.c

```diff
@@ -341,3 +342,120 @@ uint8_t has_anymod(void)
+ * \param layer layer to check against current layer.
+ * \param keycode the processed keycode, will be forwarded to handler.
+ * \param record the processed record, will be forwarded to handler.
+ * \param data anonymous data to be passed to the handler.
+ * \param handler to be called when modifiers and layer are matched.
+ * \return true if modifiers or layer did not match, return value of handler otherwise.
+ */
+bool run_on_modifiers(uint8_t modifier_mask, int16_t layer, uint16_t keycode,  keyrecord_t *record,
+                      void *data,  bool (*handler)(uint16_t keycode,  keyrecord_t *record, void *data)) {
+  static bool in_alternate_modifier;
+
+  /* nothing to do if we are not in the right layer */
+  if (layer >= 0 && !layer_state_is(layer))
+    return true;
+
+  /* when it's a key press and modifier state is pressed */
+  if (record->event.pressed && (get_mods() & modifier_mask) == modifier_mask) {
+    in_alternate_modifier = true;
+
+    /* will send modifier up so that the os won't shift the keycode we will send */
+    del_mods(modifier_mask);
+
+    /* send mods modifications */
+    send_keyboard_report();
+
+    /* execute handler */
+    bool rc = handler(keycode, record, data);
+
+    add_mods(modifier_mask);
+
+    return rc;
+  }
+
+  /* when releasing the key and we activated alternate modifier */
+  if (!record->event.pressed && in_alternate_modifier) {
+    in_alternate_modifier = false;
+
+    /* execute handler */
+    bool rc = handler(keycode, record, data);
+
+    /* make sure all mods we sat up earlier are released */
+    clear_weak_mods();
+
+    /* send mods modification */
+    send_keyboard_report();
+
+    return rc;
+  }
+
+  return true;
+}
+
+/** \brief add a weak mod if keycode has the mod enabled */
+#define maybe_add_weak_mods(keycode, mod)                               \
+  if (keycode < QK_MODS_MAX &&                                          \
+      (keycode & 0xff00) == QK_ ## mod)                                 \
+    add_weak_mods(MOD_BIT(KC_ ## mod))
+
+/** \brief copy keycode mods as weak mod */
+static void copy_mods_as_weak(uint16_t keycode) {
+    /* send mods if keycode needs it */
+    maybe_add_weak_mods(keycode, LCTL);
+    maybe_add_weak_mods(keycode, LSFT);
+    maybe_add_weak_mods(keycode, LALT);
+    maybe_add_weak_mods(keycode, LGUI);
+    maybe_add_weak_mods(keycode, RCTL);
+    maybe_add_weak_mods(keycode, RSFT);
+    maybe_add_weak_mods(keycode, RALT);
+    maybe_add_weak_mods(keycode, RGUI);
+}
+
+/** \brief run_on_modifiers handler to send custom keycode
+ */
+static bool custom_keycode_on_modifiers_handler(uint16_t keycode, keyrecord_t *record, void *data) {
+  uint16_t custom_keycode =  *((uint16_t *) data);
+
+  if (record->event.pressed) {
+    /* register custom keycode mods */
+    copy_mods_as_weak(custom_keycode);
+    send_keyboard_report();
+
+    /* send alternate key code */
+    register_code(custom_keycode);
```

Comment:
If you're allowing 16 bit integers, we should probably use the matching function here too...
```suggestion
    register_code16(custom_keycode);
```

That or limit to 8 bit keycodes.